src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,380 @@
+
+@[simp] lemma coe_zero : ↑(0 : {P : X →L[𝕜] X // is_Lprojection P}) = (0 : X →L[𝕜] X) :=
+rfl
+
+instance : has_one {P : X →L[𝕜] X // is_Lprojection P}  :=
+⟨⟨1, begin
+  rw ← sub_zero (1:X →L[𝕜] X),
+  exact is_Lprojection.Lcomplement (0 : {P : X →L[𝕜] X // is_Lprojection P}).prop,
+end⟩⟩
+
+@[simp] lemma coe_one : ↑(1 : {P : X →L[𝕜] X // is_Lprojection P}) = (1 : X →L[𝕜] X) :=
+rfl
+
+instance : bounded_order {P : X →L[𝕜] X // is_Lprojection P} :=
+{ top := 1,
+  le_top := λ P, (by rw mul_one : (↑P:X →L[𝕜] X) = ↑P * 1),
+  bot := 0,
+  bot_le := λ P, show 0 ≤ P, from zero_mul P, }
+
+@[simp] lemma coe_bot : ↑(bounded_order.bot : {P : X →L[𝕜] X // is_Lprojection P}) = (0: X →L[𝕜] X)
+  := rfl
+
+@[simp] lemma coe_top : ↑(bounded_order.top : {P : X →L[𝕜] X // is_Lprojection P}) = (1: X →L[𝕜] X)
+  := rfl
+
+lemma compl_mul_left {P : {P : X →L[𝕜] X // is_Lprojection P}} {Q: X →L[𝕜] X} :
+  Q - ↑P * Q = ↑Pᶜ * Q := by rw [coe_compl, sub_mul, one_mul]
+
+lemma compl_orthog {P : {P : X →L[𝕜] X // is_Lprojection P}} :
+  (↑P: X →L[𝕜] X) * (↑ Pᶜ) = 0 :=
+by rw [coe_compl, mul_sub, ← sq, mul_one, projection_def P.prop.left, sub_self]
+
+lemma distrib_lattice_lemma {P Q R : {P : X →L[𝕜] X // is_Lprojection P}} :
+  ((↑P:X →L[𝕜] X) + ↑Pᶜ * R) * (↑P + ↑Q * ↑R * ↑Pᶜ) = (↑P + ↑Q * ↑R * ↑Pᶜ) :=
+by rw [add_mul, mul_add, mul_add, mul_assoc ↑Pᶜ ↑R (↑Q * ↑R * ↑Pᶜ), ← mul_assoc ↑R (↑Q*↑R)  ↑Pᶜ,
+    ← coe_inf Q, commute.eq (Lproj_commute Pᶜ.prop R.prop),
+    commute.eq (Lproj_commute (Q⊓R).prop Pᶜ.prop), commute.eq (Lproj_commute R.prop (Q⊓R).prop),
+    coe_inf Q, mul_assoc ↑Q, ← mul_assoc, mul_assoc ↑R, commute.eq (Lproj_commute Pᶜ.prop P.prop),
+    compl_orthog, zero_mul, mul_zero, zero_add, add_zero, ← mul_assoc, ← sq, ← sq,
+    projection_def P.prop.left, projection_def R.prop.left, ← coe_inf Q, mul_assoc,
+    commute.eq (Lproj_commute (Q⊓R).prop Pᶜ.prop), ← mul_assoc, ← sq, projection_def Pᶜ.prop.left]
+
+instance : distrib_lattice {P : X →L[𝕜] X // is_Lprojection P} :=
+{ le_sup_left := λ P Q, (by rw [coe_sup, ← add_sub, mul_add, mul_sub, ← mul_assoc, ← sq,
+      projection_def P.prop.left, sub_self, add_zero] : ↑P = ↑P * ↑(P ⊔ Q)),
+  le_sup_right := λ P Q, by exact
+    ( begin
+      rw [coe_sup, ← add_sub, mul_add, mul_sub, commute.eq (Lproj_commute P.prop Q.prop),
+        ← mul_assoc, ← sq, projection_def Q.prop.left],
+      abel,
+    end : (↑Q: X →L[𝕜] X) = ↑Q * ↑(P ⊔ Q)),
+  sup_le := λ P Q R,
+  begin
+    intros h₁ h₂,
+    have e₁: ↑P = ↑P * ↑R := h₁,
+    have e₂: ↑Q = ↑Q * ↑R := h₂,
+    exact (by rw [coe_sup, ← add_sub, add_mul, sub_mul, mul_assoc, ← e₂, ← e₁] :
+      ↑(P ⊔ Q) = ↑(P ⊔ Q) * ↑R)
+  end,
+  inf_le_left := λ P Q, (by rw [coe_inf, mul_assoc, commute.eq (Lproj_commute Q.prop P.prop),
+    ← mul_assoc, ← sq, (projection_def P.prop.left)] : ↑(P ⊓ Q) = ↑(P ⊓ Q) * ↑P),
+  inf_le_right := λ P Q, (by rw [coe_inf, mul_assoc,  ← sq, (projection_def Q.prop.left)] :
+    ↑(P ⊓ Q) = ↑(P ⊓ Q) * ↑Q),
```

Comment:
Using `le_def` avoids these type ascriptions:
```suggestion
  inf_le_left := λ P Q, (by rw [le_def, coe_inf, mul_assoc, commute.eq (Lproj_commute Q.prop P.prop),
    ← mul_assoc, ← sq, (projection_def.1 P.prop.left)]), 
  inf_le_right := λ P Q, (by rw [le_def, coe_inf, mul_assoc, ← sq, (projection_def.1 Q.prop.left)]),
```

Reply:
That's better, thanks.